init: add sub-priority to internal ordering

JordanYates opened the pull request:
Add a sub-priority field to `Z_INIT_ENTRY_SECTION`, which is used to
  ensure that multiple drivers running at the same init priority still
  run in an order that respects their devicetree dependencies.

  This is primarily useful when multiple instances of the same device are
  instantiated that can depend on each other, for example power domains
  and i2c muxes.

After this PR, an init level now runs all `SYS_INIT`/`DEVICE_DEFINE` macros first,
then `DEVICE_DT_DEFINE` macros in order of devicetree ordinals.

Linker section names now look something like this:
```
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(.z_init_POST_KERNEL?_*)))
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(.z_init_POST_KERNEL??_*)))
 .z_init_POST_KERNEL40_0006_
                0x00000000000042b8        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0017_
                0x00000000000042c0        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0018_
                0x00000000000042c8        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0019_
                0x00000000000042d0        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0020_
                0x00000000000042d8        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
```

Resolves #22545 for multiple instances of the same device

Commit:
dts: gen_defines: generate `_ORD_STR_SORTABLE`

Generate a zero padded variant of `_ORD` that is suitable for use in
linker scripts with the `SORT` property, so that `6` is correctly placed
before `24`, and so on.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
init: add sub-priority to internal ordering

Add a sub-priority field to `Z_INIT_ENTRY_SECTION`, which is used to
ensure that multiple drivers running at the same init priority still
run in an order that respects their devicetree dependencies.

This is primarily useful when multiple instances of the same device are
instantiated that can depend on each other, for example power domains
and i2c muxes.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: kernel: device: test sub-priority

Add tests for sorting devices at the same init priority based on their
devicetree ordinals.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

gmarull commented on include/zephyr/devicetree/ordinals.h:
```diff
@@ -24,6 +24,13 @@
  */
 #define DT_DEP_ORD(node_id) DT_CAT(node_id, _ORD)
 
+/**
+ * @brief Get a node's dependency ordinal in string sortable form
+ * @param node_id Node identifier
+ * @return the node's dependency ordinal as a zero-padded integer literal
+ */
+#define DT_DEP_ORD_STR_SORTABLE(node_id) DT_CAT(node_id, _ORD_STR_SORTABLE)
```
test missing?

JordanYates replied:
Basic tests added, let me know if you have anything more comprehensive in mind

Commit:
tests: lib: devicetree: api: test `DT_DEP_ORD_STR_SORTABLE`

Add basic testing of the new `DT_DEP_ORD_STR_SORTABLE` macro.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

fabiobaltieri reviewed (changes requested):
@JordanYates could you update https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities.py#L68 to parse the new format and compare/print it properly? (there's also some unit tests https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities_test.py and an integration test that somehow is still working https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities)

Commit:
scripts: build: check_init_priorities: parse new naming

Update the script to parse the new section naming. The ordering type
is converted from an integer to a tuple, which still compares correctly
due to the elementwise behaviour of tuple comparison.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
scripts: build: check_init_priorities_test: fix

Fix the test script for the changes made to section naming.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
> @JordanYates could you update https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities.py#L68 to parse the new format and compare/print it properly? (there's also some unit tests https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities_test.py and an integration test that somehow is still working https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities)

Done

Commit:
tests: misc: check_init_priorities: force two stage linking

Add `CONFIG_DEVICE_DEPS` to force two stage linking, otherwise the
following check means the test never runs:
```
if (TARGET zephyr_pre1)
  add_dependencies(zephyr_pre1 check_init_priorities_output)
endif()
```

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: misc: check_init_priorities: update output

Update the expected output of the init priority checking.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Updated output of `tests/misc/check_init_priorities` and made the check actually run.

fabiobaltieri commented on tests/misc/check_init_priorities/validate_check_init_priorities_output.py:
```diff
@@ -8,12 +8,12 @@
+        "INFO: /i2c@11112222/test-i2c-dev@12 PRE_KERNEL_1 51 31 > /i2c@11112222 PRE_KERNEL_1 50 28",
+        "ERROR: /i2c@11112222/test-i2c-dev@10 PRE_KERNEL_1 49 29 < /gpio@ffff PRE_KERNEL_1 50 27",
+        "ERROR: /i2c@11112222/test-i2c-dev@10 PRE_KERNEL_1 49 29 < /i2c@11112222 PRE_KERNEL_1 50 28",
+        "INFO: /i2c@11112222/test-i2c-dev@11 PRE_KERNEL_1 50 30 > /gpio@ffff PRE_KERNEL_1 50 27",
+        "INFO: /i2c@11112222/test-i2c-dev@11 PRE_KERNEL_1 50 30 > /i2c@11112222 PRE_KERNEL_1 50 28",
```
oh, interesting, this was a problem because matching priority meant devices would initialize in linking order. I guess that's not possible anymore with ordinals part of the priority, which is really cool. In that case matching priority I guess could be converted to error down the road.

fabiobaltieri replied:
That's really cool, I'm almost done fixing all the errors detected in CI, had no clue how to tackle warnings but this stabilizes all of them in one shot. Good stuff!

JordanYates replied:
Resolving the ordering of matching priority devices is the one and only goal of this PR, so I'm glad it does fix it :)

fabiobaltieri reviewed (approved):
Thanks for the extra fixes!

andyross reviewed (approved):
Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...

JordanYates commented:
> Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...

Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it

gmarull commented:
> > Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...
> 
> Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it

+1, the init problem® won't be easy to solve. For example, all the flexibility we've given with SYS_INIT is now hitting us hard.

fabiobaltieri commented:
> > > Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...
> > 
> > Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it
> 
> +1, the init problem® won't be easy to solve. For example, all the flexibility we've given with SYS_INIT is now hitting us hard.

I'd be surprised at this point if someone comes up with a obviously perfect solution, but for now I think these progressive changes are moving in the right direction. I'm thinking, this takes care of same-level, the build time checking does multiple-level, @bjarki-trackunit were you looking into printing the init call sequence in https://github.com/zephyrproject-rtos/zephyr/pull/55654#issuecomment-1478452472? I'm thinking that having an init sequence printer could be used by subsystem tests and downstream application to compare the final init sequence with a golden sample so that any change gets caught and manually reviewed. That'd be a pretty solid system IMO.

fabiobaltieri commented:
@JordanYates I linked this up in #release on Discord and it's been pointed out that since it's been a hot topic recently, it may be worth running it through the Architecture WG before hitting the merge button (@carlescufi). Any chance you could talk about it? Happy to help out if the time zone is a problem btw.

Tagging as DNM for now.

JordanYates commented:
I don't think the change is controversial enough to require me to personally argue its merits 🙂 1am is a high bar. If you are happy to introduce the topic that's good enough for me.

fabiobaltieri commented:
> I don't think the change is controversial enough to require me to personally argue its merits slightly_smiling_face 1am is a high bar. If you are happy to introduce the topic that's good enough for me.

Yeah well, you never know, but I think the same. Happy to present on your behalf at the next WG then, @gmarull will you be there as well?

gmarull commented:
> > I don't think the change is controversial enough to require me to personally argue its merits slightly_smiling_face 1am is a high bar. If you are happy to introduce the topic that's good enough for me.
> 
> Yeah well, you never know, but I think the same. Happy to present on your behalf at the next WG then, @gmarull will you be there as well?

I'll be there, but TBH, I'd just go ahead and merge. This is not really a substantial change, just resolves some issues present in the current implementation.

dkalowsk reviewed (approved):
Looks pretty clean!

cfriedt commented:
> > Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...
> 
> Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it

Honestly, an LTO plugin seems increasingly more appealing these days ..

carlescufi commented:
Architecture WG:

- Incremental improvement using current infrastructure. Removes dependency on linking order for devices that have the same priority. This instead appends the Devicetree ordinal to remove any dependency from linking order.
